gax/src/main/java/com/google/api/gax/batching/FlowControlEventStats.java

```diff
@@ -0,0 +1,133 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching;
+
+import static com.google.api.gax.batching.FlowController.LimitExceededBehavior;
+
+import com.google.api.core.InternalApi;
+import com.google.api.gax.batching.FlowController.FlowControlException;
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Record the statistics of flow control events.
+ *
+ * <p>This class is populated by FlowController, which will record throttling events. Currently it
+ * only keeps the last flow control event, but it could be expanded to record more information in
+ * the future. The events can be used to dynamically adjust concurrency in the client. For example:
+ *
+ * <pre>{@code
+ * // Increase flow control limits if there was throttling in the past 5 minutes and throttled time
+ * // was longer than 1 minute.
+ * while(true) {
+ *    FlowControlEvent event = flowControlEventStats.getLastFlowControlEvent();
+ *    if (event != null
+ *         && event.getTimestampMs() > System.currentMillis() - TimeUnit.MINUTES.toMillis(5)
+ *         && event.getThrottledTimeInMs() > TimeUnit.MINUTES.toMillis(1)) {
+ *      flowController.increaseThresholds(elementSteps, byteSteps);
+ *    }
+ *    Thread.sleep(TimeUnit.MINUTE.toMillis(10));
+ * }
+ * }</pre>
+ */
+@InternalApi("For google-cloud-java client use only")
+public class FlowControlEventStats {
+
+  private volatile FlowControlEvent lastFlowControlEvent;
+
+  // We only need the last event to check if there was throttling in the past X minutes, so this
+  // doesn't need to be super accurate.
+  void recordFlowControlEvent(FlowControlEvent event) {
+    if (lastFlowControlEvent == null || event.compareTo(lastFlowControlEvent) > 0) {
+      lastFlowControlEvent = event;
```

Comment:
Looks like it is the subject of concurrent execution. Please strongly consider using at least `AtomicReference` instead of `volatile` for `lastFlowControlEvent` and here do `lastFlowControlEvent.compareAndSet(old_val, new_val)` to make sure that we set this value to what it was actually compared to on the line 71.

Reply:
I dont think this is necessary...it doesnt matter which of the events gets record here. If this got throttled by multiple threads at the same millisecond, we dont actually care which event gets record.

Reply:
Hmm, keeping this as volatile for now. We could always make it `AtomicReference` later if it needs to be more accurate?